src/platform/activation_abi.cpp

```diff
@@ -0,0 +1,69 @@
+#include "pal_internal.h"
+#include "opaque_string_wrapper.h"
+#include "platform_activation.h"
+
+namespace xlang::impl
+{
+    template <typename char_type>
+    void* get_activation_factory(
+        xlang_string class_name,
+        xlang_guid const& iid)
+    {
+        for (auto current_namespace = enclosing_namespace(to_string_view<char_type>(class_name));
+            !current_namespace.empty();
+            current_namespace = enclosing_namespace(current_namespace))
+        {
+            xlang_pfn_lib_get_activation_factory pfn = try_get_activation_func(current_namespace);
+            if (pfn)
+            {
+                void* factory{};
+                xlang_result result = (*pfn)(class_name, iid, &factory);
+                if (result == xlang_error_ok)
+                {
+                    return factory;
+                }
+                else if (result != xlang_error_class_not_available)
+                {
+                    throw_result(result);
+                }
+            }
+        }
+        throw_result(xlang_error_class_not_available);
+    }
+}
+
+using namespace xlang::impl;
+
+XLANG_PAL_EXPORT xlang_result XLANG_CALL xlang_get_activation_factory(
+    xlang_string class_name,
+    xlang_guid const& iid,
+    void** factory
+) XLANG_NOEXCEPT
+try
+{
+    if (!class_name)
+    {
+        xlang::throw_result(xlang_error_invalid_arg);
+    }
+
+    auto const encoding = xlang_get_string_encoding(class_name);
```

Comment:
Looking forward to revisiting the string question now that we've made the decision to split from WinRT. This is a good example of what developers are likely to do all over the place in the hopes of optimizing string performance. _Not_ picking an encoding does not produce a developer-friendly platform.

Reply:
Agreed, this is a bit awkward, but I also feel that this layer is an outlier. Most people will be dealing with their native string type for the programming language / application, and the conversion won't be as visible. As I noted to Ryan, this layer could accept raw character strings as well since this is only called through C. It's not a projected API.

At the end of all the processing, it eventually calls LoadLibraryW, which requires at least a substring to be converted to UFF-16 anyway. On other platforms, it might be UTF-8 at the bottom. As such, I think the code (at least for Windows) should be simplified to just do processing in UTF-16, though the implementation could be templatized to prefer UTF-8 for platforms where the module loading function uses that format.

Reply:
At a minimum, a type function to convert encoding to filesystem char type "en passant" would be nice - probably many opportunities to reuse it.  I still like the idea of tailoring string conversion to both projection and component/PAL (and minimizing it), rather than just "picking one"

Reply:
Created Issue #53 it track